### PR TITLE
Apply first batch of golangci linters

### PR DESCRIPTION
**Pull request by saschagrunert**
We enable and fix the reports of the following linters:

- deadcode
- gocritic
- golint
- ineffassign
- maligned
- prealloc
- staticcheck
- unparam
- unused
- whitepsace

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -106,6 +105,7 @@ linters-settings:
       - valSwap
       - wrapperFunc
       - yodaStyleExpr
+      # - ifElseChain
```
Comment (justaugustus) [inline .golangci.yml]: Any reason we commented this one out when it was uncommented before?

File: cmd/krel/cmd/root.go [hunk 2]
```diff
@@ -57,19 +55,3 @@ func init() {
 // initConfig reads in config file and ENV variables if set.
 func initConfig() {
 }
-
-func cleanupTmpDir(dir string) error {
```
Comment (justaugustus) [inline cmd/krel/cmd/root.go]: `deadcode`?

File: cmd/krel/cmd/push.go [hunk 3]
```diff
@@ -40,18 +40,18 @@ push --bucket=kubernetes-release-$USER
...
 	dockerRegistry   string
 	extraPublishFile string
-	federation       bool
 	gcsSuffix        string
-	noUpdateLatest   bool
-	privateBucket    bool
 	releaseKind      string
 	releaseType      string
 	versionSuffix    string
+	allowDup         bool
```
Comment (justaugustus) [inline cmd/krel/cmd/push.go]: This is just sorting the types, right?

File: pkg/git/git.go [hunk 4]
```diff
@@ -335,7 +332,7 @@ func (r *Repo) MergeBase(from, to string) (string, error) {
 	commitRevs := []string{masterRef, releaseRef}
 	var res []*object.Commit
 
-	var hashes []*plumbing.Hash
+	hashes := []*plumbing.Hash{}
```
Comment (justaugustus) [inline pkg/git/git.go]: `prealloc`?

File: pkg/notes/notes.go [hunk 5]
```diff
@@ -109,48 +109,48 @@ type ReleaseNotes map[int]*ReleaseNote
 // ReleaseNotesHistory is the sorted list of PRs in the commit history
 type ReleaseNotesHistory []int
 
-// GithubApiOption is a type which allows for the expression of API configuration
+// GitHubAPIOption is a type which allows for the expression of API configuration
 // via the "functional option" pattern.
 // For more information on this pattern, see the following blog post:
 // https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis
-type GithubApiOption func(*githubApiConfig)
+type GitHubAPIOption func(*gitHubAPIConfig)
```
Comment (justaugustus) [inline pkg/notes/notes.go]: `gitHub` in `gitHubAPIConfig` feels awkward.
Was there a linter error that suggested it over `github` --> `githubAPIConfig`?

i.e., I'm aware of some guidance about maintaining proper names (prefer `GitHub` over `Github`), but it's kind of in violation anyway once it's an unexported variable.

(Same for the changes below.)

File: .golangci.yml (hunk 1 above)
Comment (saschagrunert) [reply .golangci.yml]: It reports to refactor else-if chains into switch statements, like:
https://github.com/kubernetes/release/blob/af8e773465759f580439c253259db258a56199f6/pkg/git/git.go#L111-L122

I'm not sure if we should do this and I tend to say no. But I'm open to go with the defaults.

File: cmd/krel/cmd/root.go (hunk 2 above)
Comment (saschagrunert) [reply cmd/krel/cmd/root.go]: Yes, I could also add a `nolint` if we will need it in the future. Right now the git package takes care of that cleanup.

File: cmd/krel/cmd/push.go (hunk 3 above)
Comment (saschagrunert) [reply cmd/krel/cmd/push.go]: Yes the linter is `maligned`, which is currently only able to tell us that we can safe space in the struct, but not how :see_no_evil:. But sorting the members by size gives us the best memory alignment and safes some bytes on runtime. :+1:

File: pkg/git/git.go (hunk 4 above)
Comment (saschagrunert) [reply pkg/git/git.go]: Yes, I also thought that I read somewhere that this is the more idiomatic way for golang.

File: pkg/notes/notes.go (hunk 5 above)
Comment (saschagrunert) [reply pkg/notes/notes.go]: Yep I agree, I'll change it to `githubAPIConfig`

File: cmd/krel/cmd/push.go (hunk 3 above)
Comment (justaugustus) [reply cmd/krel/cmd/push.go]: Cool!

**Comment by saschagrunert** [comment]
> Also, for future PRs, can we have one commit per linter enabled?
> e.g.,:

Yeah I thought the same after I opened the PR, sorry for that :innocent:

File: cmd/krel/cmd/root.go (hunk 2 above)
Comment (justaugustus) [reply cmd/krel/cmd/root.go]: We'll need a common cleanup function somewhere, that is separate from the git package.
So this is fine right now, but we'll need to address that in a follow-up.

File: pkg/git/git.go (hunk 4 above)
Comment (justaugustus) [reply pkg/git/git.go]: Cool! Just my curiosity; not calling out an issue or anything. :)

File: pkg/notes/notes.go (hunk 5 above)
Comment (justaugustus) [reply pkg/notes/notes.go]: Thanks!

File: .golangci.yml (hunk 1 above)
Comment (justaugustus) [reply .golangci.yml]: Let's leave that one commented out for now.
I'd say once we hit three or more chains and there's a common pattern, maybe consider `switch` statements, but I don't think we need to nit on it at the moment.

**Comment by justaugustus** [comment]
Thanks again, Sascha!
/lgtm
/approve
/hold cancel

**Pull request merged**